File: app/src/main/java/com/bryan/platform/adapter/UserFollowAdapter.java
package com.bryan.platform.adapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bryan.platform.R;
import com.bryan.platform.activity.user.UserProfileActivity;
import com.bryan.platform.databinding.ItemUserFollowBinding;
import com.bryan.platform.model.entity.User;
import com.bryan.platform.model.response.Result;
import com.bryan.platform.network.RetrofitClient;
import com.bryan.platform.network.UserFollowService;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class UserFollowAdapter extends RecyclerView.Adapter<UserFollowAdapter.UserFollowViewHolder> {

    private List<User> users = new ArrayList<>();
    private final long currentUserId;   // 当前登录用户
    private final boolean isFollowingList; // true = 关注列表，false = 粉丝列表

    private final UserFollowService service =
            RetrofitClient.getInstance().createService(UserFollowService.class);

    // 外部回调：关注状态变化
    public interface OnFollowStatusChanged {
        void onChanged();
    }

    private OnFollowStatusChanged listener;

    public void setOnFollowStatusChanged(OnFollowStatusChanged l) {
        this.listener = l;
    }

    public UserFollowAdapter(List<User> users, long currentUserId, boolean isFollowingList) {
        this.users = users != null ? users : new ArrayList<>();
        this.currentUserId = currentUserId;
        this.isFollowingList = isFollowingList;
    }

    @NonNull
    @Override
    public UserFollowViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemUserFollowBinding binding = ItemUserFollowBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false);
        return new UserFollowViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull UserFollowViewHolder holder, int position) {
        holder.bind(users.get(position));
    }

    @Override
    public int getItemCount() {
        return users.size();
    }

    @SuppressLint("NotifyDataSetChanged")
    public void updateData(List<User> newUsers) {
        users.clear();
        if (newUsers != null) users.addAll(newUsers);
        notifyDataSetChanged();
    }

    /* ========= ViewHolder ========= */
    class UserFollowViewHolder extends RecyclerView.ViewHolder {
        private final ItemUserFollowBinding binding;

        UserFollowViewHolder(@NonNull ItemUserFollowBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(User user) {
            binding.tvUserName.setText(user.getUsername());
            // TODO: 加载头像 binding.ivUserAvatar.setImageResource(...)

            Button btn = binding.btnFollowStatus;
            if (isFollowingList) {
                setUnfollowButton(btn);
            } else {
                checkFollowStatus(user.getId(), btn);
            }

            // 关注/取消关注点击
            btn.setOnClickListener(v -> {
                if (btn.getText().toString().equals(v.getContext().getString(R.string.unfollow_button))) {
                    unfollowUser(user.getId(), btn);
                } else {
                    followUser(user.getId(), btn);
                }
            });

            // 点击整行进入用户主页
            itemView.setOnClickListener(v -> {
                Context ctx = v.getContext();
                Intent intent = new Intent(ctx, UserProfileActivity.class);
                intent.putExtra("userId", user.getId());
                ctx.startActivity(intent);
            });
        }

        /* ---------- 网络请求 ---------- */

        private void checkFollowStatus(long targetUserId, Button btn) {
            service.isFollowing(targetUserId).enqueue(new Callback<Result<Boolean>>() {
                @Override
                public void onResponse(@NonNull Call<Result<Boolean>> call,
                                       @NonNull Response<Result<Boolean>> response) {
                    if (response.isSuccessful() && response.body() != null
                            && Boolean.TRUE.equals(response.body().isSuccess())
                            && Boolean.TRUE.equals(response.body().getData())) {
                        setUnfollowButton(btn);
                    } else {
                        setFollowButton(btn);
                    }
                }

                @Override
                public void onFailure(@NonNull Call<Result<Boolean>> call, @NonNull Throwable t) {
                    setFollowButton(btn);
                    Toast.makeText(btn.getContext(), "检查关注状态失败", Toast.LENGTH_SHORT).show();
                }
            });
        }

        private void followUser(long targetUserId, Button btn) {
            btn.setEnabled(false);
            service.followUser(targetUserId).enqueue(new Callback<Result<Boolean>>() {
                @Override
                public void onResponse(@NonNull Call<Result<Boolean>> call,
                                       @NonNull Response<Result<Boolean>> response) {
                    btn.setEnabled(true);
                    if (response.isSuccessful() && response.body() != null
                            && Boolean.TRUE.equals(response.body().isSuccess())
                            && Boolean.TRUE.equals(response.body().getData())) {
                        setUnfollowButton(btn);
                        Toast.makeText(btn.getContext(), "关注成功", Toast.LENGTH_SHORT).show();
                        if (listener != null) listener.onChanged();
                    } else {
                        String msg = response.body() != null ? response.body().getMessage() : "关注失败";
                        Toast.makeText(btn.getContext(), msg, Toast.LENGTH_SHORT).show();
                    }
                }

                @Override
                public void onFailure(@NonNull Call<Result<Boolean>> call, @NonNull Throwable t) {
                    btn.setEnabled(true);
                    Toast.makeText(btn.getContext(), "网络错误: " + t.getMessage(), Toast.LENGTH_SHORT).show();
                }
            });
        }

        private void unfollowUser(long targetUserId, Button btn) {
            btn.setEnabled(false);
            service.unfollowUser(targetUserId).enqueue(new Callback<Result<Boolean>>() {
                @Override
                public void onResponse(@NonNull Call<Result<Boolean>> call,
                                       @NonNull Response<Result<Boolean>> response) {
                    btn.setEnabled(true);
                    if (response.isSuccessful() && response.body() != null
                            && Boolean.TRUE.equals(response.body().isSuccess())
                            && Boolean.TRUE.equals(response.body().getData())) {
                        setFollowButton(btn);
                        Toast.makeText(btn.getContext(), "取消关注成功", Toast.LENGTH_SHORT).show();
                        if (listener != null) listener.onChanged();
                    } else {
                        String msg = response.body() != null ? response.body().getMessage() : "取消关注失败";
                        Toast.makeText(btn.getContext(), msg, Toast.LENGTH_SHORT).show();
                    }
                }

                @Override
                public void onFailure(@NonNull Call<Result<Boolean>> call, @NonNull Throwable t) {
                    btn.setEnabled(true);
                    Toast.makeText(btn.getContext(), "网络错误: " + t.getMessage(), Toast.LENGTH_SHORT).show();
                }
            });
        }

        /* ---------- UI 样式 ---------- */

        private void setFollowButton(Button btn) {
            btn.setText(btn.getContext().getString(R.string.follow_button));
            btn.setBackgroundResource(R.drawable.btn_follow_background);
            btn.setTextColor(btn.getContext().getColor(R.color.white));
        }

        private void setUnfollowButton(Button btn) {
            btn.setText(btn.getContext().getString(R.string.unfollow_button));
            btn.setBackgroundResource(R.drawable.btn_unfollow_background);
            btn.setTextColor(btn.getContext().getColor(R.color.black));
        }
    }
}

File: app/src/main/java/com/bryan/platform/adapter/UserFollowAdapter.kt
package com.bryan.platform.adapter

import android.annotation.SuppressLint
import android.content.Intent
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Button
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.bryan.platform.R
import com.bryan.platform.activity.user.UserProfileActivity
import com.bryan.platform.databinding.ItemUserFollowBinding
import com.bryan.platform.model.entity.User
import com.bryan.platform.model.response.Result
import com.bryan.platform.network.RetrofitClient
import com.bryan.platform.network.UserFollowService
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class UserFollowAdapter(
    private var users: List<User>,
    private val currentUserId: Long, // Current logged-in user's ID
    private val isFollowingList: Boolean // true if this adapter is for 'following' list, false for 'followers'
) : RecyclerView.Adapter<UserFollowAdapter.UserFollowViewHolder>() {

    private val userFollowService: UserFollowService = RetrofitClient.createService(UserFollowService::class.java)

    // Callback to update the list in the Activity after a follow/unfollow action
    var onFollowStatusChanged: (() -> Unit)? = null

    inner class UserFollowViewHolder(private val binding: ItemUserFollowBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(user: User) {
            binding.tvUserName.text = user.username

            // You can add logic here to load user avatars if available
            // binding.ivUserAvatar.setImageResource(R.drawable.ic_default_profile)

            // Determine initial follow status and update button text/style
            // For following list, the button is always "Unfollow"
            // For followers list, we need to check if current user is following this follower
            if (isFollowingList) {
                setUnfollowButton(binding.btnFollowStatus)
            } else {
                checkFollowStatus(currentUserId, user.id, binding.btnFollowStatus)
            }

            binding.btnFollowStatus.setOnClickListener {
                if (binding.btnFollowStatus.text == itemView.context.getString(R.string.unfollow_button)) {
                    unfollowUser(currentUserId, user.id, binding.btnFollowStatus)
                } else {
                    followUser(currentUserId, user.id, binding.btnFollowStatus)
                }
            }

            itemView.setOnClickListener {
                val context = itemView.context
                val intent = Intent(context, UserProfileActivity::class.java)
                intent.putExtra("userId", user.id)
                context.startActivity(intent)
            }

        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserFollowViewHolder {
        val binding = ItemUserFollowBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return UserFollowViewHolder(binding)
    }

    override fun onBindViewHolder(holder: UserFollowViewHolder, position: Int) {
        holder.bind(users[position])
    }

    override fun getItemCount(): Int = users.size

    @SuppressLint("NotifyDataSetChanged")
    fun updateData(newUsers: List<User>) {
        users = newUsers
        notifyDataSetChanged() // For simplicity, re-drawing the whole list
    }

    private fun checkFollowStatus(followerId: Long, followingId: Long, button: Button) {
        userFollowService.isFollowing(followingId).enqueue(object : Callback<Result<Boolean>> {
            override fun onResponse(call: Call<Result<Boolean>>, response: Response<Result<Boolean>>) {
                if (response.isSuccessful && response.body()?.isSuccess() == true) {
                    val isFollowing = response.body()?.data ?: false
                    if (isFollowing) {
                        setUnfollowButton(button)
                    } else {
                        setFollowButton(button)
                    }
                } else {
                    // Handle error, maybe set button to a default state or disable
                    setFollowButton(button) // Default to follow if status check fails
                    Toast.makeText(button.context, "检查关注状态失败", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<Result<Boolean>>, t: Throwable) {
                setFollowButton(button) // Default to follow on network error
                Toast.makeText(button.context, "网络错误: ${t.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun followUser(followerId: Long, followingId: Long, button: Button) {
        button.isEnabled = false // Disable button to prevent multiple clicks
        userFollowService.followUser(followingId).enqueue(object : Callback<Result<Boolean>> {
            override fun onResponse(call: Call<Result<Boolean>>, response: Response<Result<Boolean>>) {
                button.isEnabled = true
                if (response.isSuccessful && response.body()?.isSuccess() == true && response.body()?.data == true) {
                    setUnfollowButton(button)
                    Toast.makeText(button.context, "关注成功", Toast.LENGTH_SHORT).show()
                    onFollowStatusChanged?.invoke() // Notify activity to refresh counts
                } else {
                    val errorMsg = response.body()?.message ?: "关注失败"
                    Toast.makeText(button.context, errorMsg, Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<Result<Boolean>>, t: Throwable) {
                button.isEnabled = true
                Toast.makeText(button.context, "网络错误: ${t.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun unfollowUser(followerId: Long, followingId: Long, button: Button) {
        button.isEnabled = false // Disable button to prevent multiple clicks
        userFollowService.unfollowUser(followingId).enqueue(object : Callback<Result<Boolean>> {
            override fun onResponse(call: Call<Result<Boolean>>, response: Response<Result<Boolean>>) {
                button.isEnabled = true
                if (response.isSuccessful && response.body()?.isSuccess() == true && response.body()?.data == true) {
                    setFollowButton(button)
                    Toast.makeText(button.context, "取消关注成功", Toast.LENGTH_SHORT).show()
                    onFollowStatusChanged?.invoke() // Notify activity to refresh counts
                } else {
                    val errorMsg = response.body()?.message ?: "取消关注失败"
                    Toast.makeText(button.context, errorMsg, Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<Result<Boolean>>, t: Throwable) {
                button.isEnabled = true
                Toast.makeText(button.context, "网络错误: ${t.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun setFollowButton(button: Button) {
        button.text = button.context.getString(R.string.follow_button)
        button.setBackgroundResource(R.drawable.btn_follow_background) // Use a custom drawable for follow button
        button.setTextColor(button.context.getColor(R.color.white))
    }

    private fun setUnfollowButton(button: Button) {
        button.text = button.context.getString(R.string.unfollow_button)
        button.setBackgroundResource(R.drawable.btn_unfollow_background) // Use a custom drawable for unfollow button
        button.setTextColor(button.context.getColor(R.color.black))
    }
}
